apps/openmw/mwworld/containerstore.hpp

```diff
@@ -187,96 +200,143 @@ namespace MWWorld
+        mutable PtrType mPtr;
+
+        typename Iterator<ESM::Potion>::type mPotion;
+        typename Iterator<ESM::Apparatus>::type mApparatus;
+        typename Iterator<ESM::Armor>::type mArmor;
+        typename Iterator<ESM::Book>::type mBook;
+        typename Iterator<ESM::Clothing>::type mClothing;
+        typename Iterator<ESM::Ingredient>::type mIngredient;
+        typename Iterator<ESM::Light>::type mLight;
+        typename Iterator<ESM::Lockpick>::type mLockpick;
+        typename Iterator<ESM::Miscellaneous>::type mMiscellaneous;
+        typename Iterator<ESM::Probe>::type mProbe;
+        typename Iterator<ESM::Repair>::type mRepair;
+        typename Iterator<ESM::Weapon>::type mWeapon;
+
+        ContainerStoreIteratorBase (ContainerStoreType container);
+        ///< End-iterator
+
+        ContainerStoreIteratorBase (int mask, ContainerStoreType container);
+        ///< Begin-iterator
+
+        // construct iterator using a CellRefList iterator
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Potion>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Apparatus>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Armor>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Book>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Clothing>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Ingredient>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Light>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Lockpick>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Miscellaneous>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Probe>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Repair>::type);
+        ContainerStoreIteratorBase (ContainerStoreType container, typename Iterator<ESM::Weapon>::type);
+
+        template<class T>
+        void copy (const ContainerStoreIteratorBase<T>& src);
+        
+        void incType ();
+        
+        void nextType ();
+
+        bool resetIterator ();
+        ///< Reset iterator for selected type.
+        ///
+        /// \return Type not empty?
+
+        bool incIterator ();
+        ///< Increment iterator for selected type.
+        ///
+        /// \return reached the end?
 
         public:
+            template<class T>
+            ContainerStoreIteratorBase (const ContainerStoreIteratorBase<T>& other)
+            {
+                char CANNOT_CONVERT_CONST_ITERATOR_TO_ITERATOR[IsConvertible<T, PtrType, void>::value ? 1 : -1];
```

Comment:
@scrawl Got the linking error issue solved with this.

Error for this code:
```c++
// I didn't use cbegin() to make sure we can convert from nonconst to const
// and cend() to make sure we can compare const with nonconst
for (MWWorld::ConstContainerStoreIterator it = invStore.begin(); it != invStore.end(); ++it)
{
    MWWorld::ContainerStoreIterator nonconst = it; // error
    /* ... */
}
```
GCC:
![gcc_error](https://cloud.githubusercontent.com/assets/11935784/23097003/1e9bc3fa-f621-11e6-8162-ef524e6c02e2.png)

CLANG:
![clang_error](https://cloud.githubusercontent.com/assets/11935784/23097069/8c152c40-f622-11e6-98ba-b1b5a699880d.png)